### ISSUE TITLE
Undeterministic custom keymap order?

### ISSUE THREAD

**Issue by ywpkwon** [AstroNvim/AstroNvim#1222]
### Checklist

- [X] Using a stable version of Neovim (i.e. not neovim nightly)
- [X] `:PackerSync`
- [X] `:AstroUpdate`
- [X] Restarted AstroNvim

### AstroNvim version

Version: v2.5.0

### Neovim version (>= 0.8)

NVIM v0.8.0-1210-gd367ed9b2

### Operating system/version

Ubuntu 22.04

### Describe the bug

I wanted to override `H` and `L` as `beginning of line` and `end of line` from the original `previous buffer` and `next buffer`, respectively. But I think there's randomness which one (user config vs original config) runs later.

When I launch nvim, 
- sometime H and L works as I want. 
- (restart) sometime H and L shows original behavior.
- (restart) sometime I also verified H does `previous buffer`, but L does `end of line`. (i.e., configuration mixed) ... :open_mouth: !?

So I think there's randomeness in which is called later. Can you check regard to this matter? What can affect this?

```lua
-- lua/user/init.lua
  mappings = {
    n = {
      -- second key is the lefthand side of the map
      -- mappings seen under group name "Buffer"
      -- ["<leader>bb"] = { "<cmd>tabnew<cr>", desc = "New tab" },
      -- ["<leader>bc"] = { "<cmd>BufferLinePickClose<cr>", desc = "Pick to close" },

      [";"] = { ":", desc = ":" }, -- let's try to switch ; and :
      [":"] = { ";", desc = ";" }, -- let's try to switch ; and :
      ["H"] = { "^", desc = "beginning of line" },
      ["L"] = { "$", desc = "end of line" },
    },
    v = {
      [";"] = { ":", desc = ":" }, -- let's try to switch ; and :
      [":"] = { ";", desc = ";" }, -- let's try to switch ; and :
      ["H"] = { "^", desc = "beginning of line" },
      ["L"] = { "$", desc = "end of line" },
    },
```

* FYI, non-overriding custom key mappings seems to work great.
* I asked this question to the closed #1154 a week ago, there was no response so I created a new one.

### Steps to Reproduce

```lua
-- lua/user/init.lua

  mappings = {
    n = {
      -- second key is the lefthand side of the map
      -- mappings seen under group name "Buffer"
      -- ["<leader>bb"] = { "<cmd>tabnew<cr>", desc = "New tab" },
      -- ["<leader>bc"] = { "<cmd>BufferLinePickClose<cr>", desc = "Pick to close" },

      [";"] = { ":", desc = ":" }, -- let's try to switch ; and :
      [":"] = { ";", desc = ";" }, -- let's try to switch ; and :
      ["H"] = { "^", desc = "beginning of line" },
      ["L"] = { "$", desc = "end of line" },
    },
    v = {
      [";"] = { ":", desc = ":" }, -- let's try to switch ; and :
      [":"] = { ";", desc = ";" }, -- let's try to switch ; and :
      ["H"] = { "^", desc = "beginning of line" },
      ["L"] = { "$", desc = "end of line" },
    },
```

### Expected behavior

always overriding original behavior with the user keymap.

**Comment by mehalter** [comment]
The way the `mappings` table works is the key lets you completely overwrite the original configuration. The key is that you need to check for original keymaps for what key we are using. So for `H` and `L` we are actually using `<S-h>` and `<S-l>` rather than `H` and `L`. With that in mind, try this:

```lua
  mappings = {
    n = {
      -- second key is the lefthand side of the map
      -- mappings seen under group name "Buffer"
      -- ["<leader>bb"] = { "<cmd>tabnew<cr>", desc = "New tab" },
      -- ["<leader>bc"] = { "<cmd>BufferLinePickClose<cr>", desc = "Pick to close" },

      [";"] = { ":", desc = ":" }, -- let's try to switch ; and :
      [":"] = { ";", desc = ";" }, -- let's try to switch ; and :
      ["<S-h>"] = { "^", desc = "beginning of line" },
      ["<S-l>"] = { "$", desc = "end of line" },
    },
    v = {
      [";"] = { ":", desc = ":" }, -- let's try to switch ; and :
      [":"] = { ";", desc = ";" }, -- let's try to switch ; and :
      ["<S-h>"] = { "^", desc = "beginning of line" },
      ["<S-l>"] = { "$", desc = "end of line" },
    },
  }
```

This will make it so our original bindings never get set in the first place. Sorry this is a bit annoying and takes some forethought before making mappings :( I'm not sure of a good way to generalize the normalization of keymaps in a way to make these things interchangeable.

**Comment by ywpkwon** [comment]
Thanks for your explanation. I'm slightly confusing here. 
> The way the mappings table works is the key lets you completely overwrite the original configuration.

Could you please verify my understanding? Let's say
- I don't need original `H` and `L` behavior. (buffer navigation) 
- I just want to set `H` and `L` in my way.

With `mappings` table, I thought I can overwrite the original configuration. But you mean, 
- There is no easy way to overwrite original configuration.
- I rather have to avoid existing ones and introduce unbind key like `S-h` and `S-l`

Is my understanding correct?

**Comment by mehalter** [comment]
You can definitely overwrite the original configuration easily :) You can instead unmap `<S-h>` and `<S-l>` if you want and map `H` and `L` directly. Something like this:

```lua


  mappings = {
    n = {
      -- unmap original keys
      ["<S-h>"] = false,
      ["<S-l>"] = false,
      -- second key is the lefthand side of the map
      -- mappings seen under group name "Buffer"
      -- ["<leader>bb"] = { "<cmd>tabnew<cr>", desc = "New tab" },
      -- ["<leader>bc"] = { "<cmd>BufferLinePickClose<cr>", desc = "Pick to close" },

      [";"] = { ":", desc = ":" }, -- let's try to switch ; and :
      [":"] = { ";", desc = ";" }, -- let's try to switch ; and :
      H = { "^", desc = "beginning of line" },
      L = { "$", desc = "end of line" },
    },
    v = {
      [";"] = { ":", desc = ":" }, -- let's try to switch ; and :
      [":"] = { ";", desc = ";" }, -- let's try to switch ; and :
      H = { "^", desc = "beginning of line" },
      L = { "$", desc = "end of line" },
    },
  }
```

You can unmap any default mapping by setting the key to `false`. Hopefully that answers your question!

Also I don't quite understand what you mean by "There is no easy way to overwrite original configuration". The `mappings` table directly overwrites the original configuration table before it is used to set the mappings. So if you mapped it to `<S-h>` and `<S-l>` it will overwrite the defaults for that. Or instead you can set those to `false` so that they never get set in the first place and you can map `H` and `L` as you would like.

**Comment by ywpkwon** [comment]
Oh! I got the meaning. This is why I was confusing. 

- I've never thought there is distinction between the capital letter `H` and the `<S-h>` (shift + h)
- At first I thought you are saying "instead use a control modifer like `<C-h>` or something." Now I realized its the shift key. :)
- So I was like, "saying possible but why recommend trying control key"? 

Thanks again for your explanation. :pray:  It works well. I'm fine just with `<S-h>` and `<S-l>`.
